smart_open/tests/test_azure.py

```diff
@@ -618,6 +619,66 @@ def test_write_blob_client(self):
         ))
         self.assertEqual(output, [test_string])
 
+    def test_abort_upload(self):
+        """Does aborted upload skip commit_block_list?"""
+        test_string = b"42" * 42
+        blob_name = "test_abort_upload_%s" % BLOB_NAME
+
+        container_client = CLIENT.get_container_client(CONTAINER_NAME)
+        blob_client = container_client.get_blob_client(blob_name)
+
+        try:
+            with smart_open.open(
+                "azure://%s/%s" % (CONTAINER_NAME, blob_name),
+                "wb",
+                transport_params={"client": blob_client, "min_part_size": 42},
+            ) as fout:
+                fout.write(test_string)
+                raise ValueError
+        except ValueError:
+            # FakeBlobClient.commit_block_list was not called
+            self.assertGreater(len(blob_client._staged_contents), 0)
+
+    def test_abort_upload_text_mode(self):
+        """Does aborted upload skip commit_block_list in text mode?"""
+        test_string = "42" * 42
+        blob_name = "test_abort_upload_%s" % BLOB_NAME
+
+        container_client = CLIENT.get_container_client(CONTAINER_NAME)
+        blob_client = container_client.get_blob_client(blob_name)
+
+        try:
+            with smart_open.open(
+                "azure://%s/%s" % (CONTAINER_NAME, blob_name),
+                "w",
+                transport_params={"client": blob_client, "min_part_size": 42},
+            ) as fout:
+                fout.write(test_string)
+                raise ValueError
+        except ValueError:
+            # FakeBlobClient.commit_block_list was not called
+            self.assertGreater(len(blob_client._staged_contents), 0)
```

Comment:
this one still fails, because this test case covers the TextIOWrapper case [which calls `close`](https://github.com/python/cpython/blob/v3.11.1/Lib/_pyio.py#L2194) on the azure binary io before our `FileLikeProxy` had the chance to call `__exit__` on the azure binary io (which circumvents `close` when `__exit__` is called during exception handling ref https://github.com/piskvorky/smart_open/pull/786)

Reply:
471b4b3